[Runtime/ABI] Replace `emitTypeFieldAccessor` with a single runtime method

xedin opened the pull request:
Emitting individual accessors per field for each nominal type is inefficient from code size
and general code reusability perspective, because that duplicates data which already
exists in reflection section of the binary. This PR is a series of changes to get rid of
special lazy accessors and replace them with a single runtime method to request type field
metadata.

Resolves: rdar://problem/26060144

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[Runtime] Register type field metadata section

xedin commented:
@DougGregor @jckarter @slavapestov Can you please take a look if this is a a correct initial direction?

slavapestov commented on include/swift/Runtime/RuntimeFunctions.def:
```diff
@@ -1222,6 +1222,14 @@ FUNCTION(RegisterTypeMetadataRecords,
 
+// void swift_registerTypeFields(const FieldDescriptor *begin,
+//                               const FieldDescriptor *end)
+FUNCTION(RegisterTypeFields,
```
This file is only used for functions called from IRGen, so I don't think this is necessary.

xedin replied:
Gotcha! Will remove.

slavapestov commented on stdlib/public/runtime/MetadataLookup.cpp:
```diff
@@ -313,6 +311,56 @@ _findProtocolDescriptor(llvm::StringRef mangledName) {
+
+struct TypeFieldMetadataState {
+  ConcurrentMap<DescriptorCacheEntry<FieldDescriptor>> FieldCache;
+  std::vector<FieldMetadataSection> SectionsToScan;
```
Are we allowed to use std::vector from the runtime?

xedin replied:
That's what the rest of the caches in this file are doing too - protocols and type metadata, is that incorrect?

DougGregor replied:
`std::vector` is okay in the runtime.

slavapestov commented:
This looks like a good start, I'm looking forward to this coming together!

xedin committed:
[Runtime/Reflection] Add `swift_getFieldAt` runtime accessor

Use information from reflection section of the binary to lookup
type field info such as name and it's type and return it using
new `swift_getFieldAt` method based on nominal type and field index.

xedin commented:
I've started using `swift_getTypeByMangledName` but I have no idea what were to get arguments for it expect for the mangled name, is that ok?

Also I've noticed that reflection field record doesn't have "weak" in flags, is that intentional?

DougGregor commented on stdlib/public/runtime/MetadataLookup.cpp:
```diff
@@ -748,3 +791,39 @@ swift_getTypeByMangledName(const char *typeNameStart, size_t typeNameLength,
+    auto type = field.getMangledTypeName(0);
+
+    auto *typeMetadata = swift_getTypeByMangledName(type.data(), type.length(),
+                                                    0, nullptr, nullptr);
```
Until we're able to provide the substitutions to `swift_getTypeByMangledName`, this won't work for generic types. There's no good way to get all of those substitutions correctly now (the information isn't there in the nominal type descriptors), but you should be able to handle non-nested generics with no requirements by using `TargetMetadata::getGenericArgs()`. That would at least be enough to see basic substitutions working.

Fully supporting substitutions here will require @jckarter  to land https://github.com/apple/swift/pull/13468

xedin replied:
Good to know, thanks, @DougGregor ! In a state I have it right now, would it be able to find non-generic types? Also, all of the methods in `Field{Descriptor, Record}` require `Offset` as an argument, I looked at how `swift-reflection-dump` gets it, but I don't think it's possible to do the same in `ImageInspection{ELF, MachO}` because we there is no `SectionRef`, is there some other way to get it without using object file interface?

xedin replied:
Talked to @jckarter offline and it looks like I don't need to worry about `offset` I mentioned in my previous comment. So the only thing which remains is to get some information from given `type` to supply generic env to the `swift_getTypeByMangledName` call and wait for other changes to be landed.

DougGregor reviewed (commented):
Definitely looks like the right approach!

xedin commented on stdlib/public/runtime/MetadataLookup.cpp:
```diff
@@ -755,3 +798,55 @@ swift_getTypeByMangledName(const char *typeNameStart, size_t typeNameLength,
+
+    auto *typeMetadata =
+        swift_getTypeByMangledName(type.data(), type.length(), numberOfLevels,
+                                   paramsPerLevel, base->getGenericArgs());
```
@DougGregor Is this the correct way of using `getGenericArgs`? I've also modified interface a bit to use a callback instead of trying to return `std::pair`, WDYT?

jckarter replied:
Since you're using a callback, you might also be able to make the callback use `StringRef` instead of `std::string`, since it could have access to memory from `swift_getFieldAt`'s stack frame.

xedin replied:
Done!

xedin commented:
@swift-ci please smoke test

xedin commented:
Ok, I'm getting closer, only 9 stdlib tests now fail locally and I think they all related to generics, I'll continue tomorrow.

xedin commented:
@DougGregor @jckarter I got this to the point where only failing tests are related to unimplemented things in `DecodedMetadataBuilder` like weak storage.

xedin committed:
[Runtime/ABI] Add ownership info to the name-to-type metadata builder

xedin commented:
Only remaining problems in tests are related to REPL, so I'm trying to figure out how to register new types there. It seems like newly defined types are not picked up automatically via image inspection callbacks.

jckarter commented:
@xedin Awesome! In JIT mode, we need to manually emit a call that registers the data from the newly-generated code since there's no "image" per se. Look at `IRGenModule::emitRuntimeRegistration()`, which generates runtime calls to manually register generated code with the ObjC and Swift runtimes in JIT mode.

xedin committed:
[Runtime/Metadata] Add support for dynamic field descriptor registration

xedin commented:
@DougGregor dynamic registration now works! I'm waiting for https://github.com/apple/swift/pull/14038 to go in before all of the tests are green (I think), because it fails to demangle some imported types in function counter tests.

xedin commented:
Ok I verified that we can't produce metadata for 'Builtin.RawPointer' and other 'Builtin' types right now, that's why function counter tests are failing.

slavapestov commented:
This is looking good! Don't forget to remove the old mechanism, too.

slavapestov commented:
For weak and unowned properties we need to change @DougGregor's entry point a bit. Since 'weak T' and 'unowned T' don't have type metadata distinct from T, the entry point should just unwrap the storage qualifier before building metadata. However, it needs to return a flags word, in addition to the type metadata, which indicates if the original mangled type was weak or unowned.

xedin commented:
Thanks, @slavapestov! Will remove the old code once tests are all green! :)

xedin committed:
[Runtime/Reflection] Switch enum mirrors to use runtime get field method

xedin commented:
In my local build the biggest difference is `libSwiftCore` (which I think should be stdlib with everything else?), it's almost 64KB smaller than before.

xedin commented:
Rebased since @DougGregor changes for imported types went in and this is down to 2 failures - only function counter ones, which reference built-in types...

xedin committed:
[Runtime] NFC: Update names of the field section to swift5_fieldmd

xedin committed:
[IRGen] Don't generate field reflection metadata for opaque struct/enum

One of my previous commits did that, but since we have Builtins reflection
section we don't really want to generate reflection field metadata for
opaque structs/enums. This reverts to the previous behavior.

slavapestov commented on include/swift/Runtime/Metadata.h:
```diff
@@ -1233,17 +1230,14 @@ struct TargetNominalTypeDescriptor final
-      /// length and order is consistent with that of the field offset vector.
+
+    private:
+      // FIXME: This is left as a placeholder (which is always 0) because
```
I thought GenReflection.cpp emits enum case names in the field names section too. The runtime mirrors implementation should use that, and we should remove this field entirely.

jckarter replied:
I think we can change the NominalTypeDescriptor layout as a separate step once the implementation no longer needs these fields. Pavel tried removing the fields and ran into other problems where the layout was depended on. Hopefully I shook those out since I just changed the format of everything in #13468, but it still doesn't hurt to make incremental steps.

xedin committed:
[Runtime/Metadata] Move `swift_getFieldAt` to new generic context framework

xedin committed:
[Metadata] Use ownership info from _typeByMangledName

xedin committed:
[MetadataLookup] Properly setup Demangler symbolic ref resolver

jckarter committed:
Demangler: Fill in places we need to handle symbolic references and OtherNominalTypes when demangling bound generic types.

xedin committed:
[IRGen/Reflection] Emit built-in reflection metadata in JIT mode

Since Mirrors are using swift_getFieldAt we need built-in type
metadata present in JIT/REPL mode as well, otherwise some of the
types are going to become unprintable.

xedin committed:
[IRGen] Remove lazy field type accessor functions

All of their usages have been replaced with new runtime
`swift_getFieldAt` method.

xedin committed:
[IRGen] Remove 'FieldNames' field from type context descriptor

All of the information contained by this field (list of property names)
is already encoded as part of the field reflection metadata and
is accessible via `swift_getFieldAt` runtime method.

xedin commented:
/cc @jckarter @DougGregor I was able to remove both field accessors and names from the context descriptor too!

xedin commented:
@swift-ci please test

xedin commented:
Interesting, looks like only on `watchsimilator-i386` tests `TupleMirror/NoLeak` failed in StringRef. I'll try to re-run to see if it has to do with the bot because that test was successful on all other platforms and I didn't modify anything in the TupleMirror code.

xedin commented:
@swift-ci please clean test macOS platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@jckarter Any idea what this test might be only failing on watchsimulator-i386 with:
```
[ RUN      ] Reflection.TupleMirror/NoLeak
stderr>>> Assertion failed: (Index < Length && "Invalid index!"), function operator[], file /Users/buildnode/jenkins/workspace/swift-PR-osx/branch-master/llvm/include/llvm/ADT/StringRef.h, line 239.
stderr>>> CRASHED: SIGABRT
```

would it be related to the logic in `getMangledTypeName` ?

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please smoke benchmark compiler performance

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please smoke benchmark compiler performance

xedin commented:
@jckarter After some time of digging I could localize the problem to this snippet:

```swift
import Foundation

var nsObjectCanaryCount = 0

@objc class NSObjectCanary : NSObject {
  override init() {
    nsObjectCanaryCount += 1
  }
  deinit {
    nsObjectCanaryCount -= 1
  }
}

struct NSObjectCanaryStruct {
  var ref = NSObjectCanary()
}

func foo() {
  autoreleasepool {
    let x = (1, NSObjectCanaryStruct())
    var output = ""
    dump(x, to: &output)
    print(output)
  }
}

foo()
```

It looks like it can follow symbolic reference and get to "ref" field but after that on `watchsimulator-i386` it fails somewhere `_typeByMangledName` when fetching metadata for `NSObjectCanaryStruct`. Going to continue tomorrow

xedin commented:
Ok, I think I figured it out, shouldn't be trying to check for convenience dot syntax in mangled names if symbolic references are present.

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please test macOS platform

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please smoke test macOS platform

xedin commented:
Looks like 'smoke test' doesn't run swiftPM tests unfortunately, I'll disable crashing test and re-run whole suite...

xedin committed:
[IRGen] Ensure that all of the field type metadata is properly forced

xedin commented:
Looks like the problem related to failures in SwiftPM tests is related to removal of the field type accessors which leads to not all of the metadata being forced (when I thought it might be viable only for foreign metadata)...

xedin commented:
@swift-ci please clean test

xedin commented:
@swift-ci please clean test

xedin commented:
Switched to using `getOrCreateTypeMetadataAccessFunction` which should handled both foreign and native metadata and only special casing protocols now since they are not handled by it.

xedin commented:
@swift-ci please test

xedin commented:
@jckarter Looks like CI is finally green! I am going to run some more checks but please take a look at the code meanwhile.

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test source compatibility

xedin commented:
Ok, looks like I have to rebase changes to type context descriptors and re-run everything again :)

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please test

xedin commented:
@swift-ci please test source compatibility

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
@swift-ci please smoke test compiler performance

xedin commented:
Looks like we save about half a megabyte with this changes

jckarter commented on stdlib/public/runtime/MetadataLookup.cpp:
```diff
@@ -929,3 +1042,106 @@ swift_getTypeByMangledName(const char *typeNameStart, size_t typeNameLength,
+    getFieldAt(*Value->getDescription());
+    return;
+  }
+  */
```
Do you want to try turning the cache on before committing, or do that in a followup?

xedin replied:
I think it could be done in the follow up, I'm still trying to figure out what is safe to cache and what is not...

jckarter commented:
Looks good! It's probably a good idea to land this and then improve the layout, enable the cache, etc. in follow-up commits.

xedin commented:
@jckarter Thanks! I'll land it and do couple of improvements later then!

bitjammer commented:
I just came here to say:

THIS IS SUPERB.
